---
title: "Quantitative Metabarcoding"
author: "The Wares Lab"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: html_document
---

***n.b. maybe you should describe the pitfalls of allele frequency approaches to relative abundance, but not dwell on them. for example, though differential amplification is most certainly an issue (cite that NGS paper) there are approaches that intend to minimize this issue. additionally, a barcoding project may focus on a limited taxon, e.g. at family level or lower, in which case barcode efficiency is less problematic. here we just state that this is an issue for PCR efficiency and copy number reasons, and thus we are exploring what information is available solely from the summary statistics of pooled/aligned sequence data***

***n.b. 2 there are perhaps some advances or ideas in your Shiny paper to synchronize into this so that you and Paula can work on it.***

Documenting the distribution and abundance of biodiversity is perhaps more important now than ever as scientists evaluate how populations are responding to environmental change. Though technological advances have rapidly improved some elements of this, there are still glaring deficiencies in our ability to efficiently catalog the diversity of even related guilds at a location. Next generation sequencing has transformed the study of microbial and viral diversity, but eukaryotic diversity remains a more difficult prospect.

The reasons for this include the very diversity of these organisms. Though it is also known that "barcode" loci in microbes may vary in abundance among taxa and populations leading to inexact measures of relative abundance (picante ref), such a problem may be exacerbated when individuals of different life stages and different species vary in size and thus cellular quantity - a typical metazoan barcode locus, mitochondrial COI, would then over-represent the abundance of larger individuals. 

Additionally, differential amplification of barcode genes, with efficiency declining with every base pair mismatch between genomic and primer sites, leads to further uncontrolled biases in enumerating the relative abundance of taxa (DOI: 10.1111/nph.12923) DOI: 10.1111/1755-0998.12355. Given the goal of including diversity that may be distantly related (e.g. the zooplankton from freshwater pools in South Carolina incorporate greater than 400 million years of phylogenetic diversity), we are often left with presence-absence data at best unless a skilled taxonomist can individually key these organisms out.

The approach in barcoding individuals of any taxon has typically followed a pattern of exploring the sequence divergence of samples from a catalog of known diversity. Either individuals are sequenced, thus maintaining both identity and recognition of abundance - a method that has not changed in over 20 years - or samples of many individuals are sequenced, and the match of individual reads from NGS data are either used to infer relative abundance (in the case of microbial diversity) or simply used to verify the presence of particular taxa. What has not been sufficiently considered is the baseline information we may have about a population or species that is measurable with these same data, in terms of genetic diversity.

The summary statistics for DNA sequence diversity are well established and generally recognize the population mutation rate THETA at a given locus; as a population increases in size, or as the mutation rate at that locus increases, more polymorphisms and more diversity will be found. There are limitations to this approach based on Kimura's neutral theory, as various forms of genomic selection will limit the direct relationship between population size and population diversity (Wares 2010). Nevertheless, these summary statistics - including Watterson's theta, a sample-normalized estimator of THETA using the number of segregating sites in a sample - may provide information necessary to generate abundance patterns from NGS data. However, this information also has its limits: nucleotide diversity (pi) will be biased by differential amplification across individuals, as well as relatively uninformative - or diminishing returns - as the number of sampled individuals increases (Wakeley 2008). Haplotype diversity (H) is likely sufficient to set a minimum boundary on the number of individuals sampled, and H along with S have some information about the probability associated with larger numbers of individuals.

Here we present the mathematical considerations necessary to develop these quantitative tools, and then apply the method to data from a well-characterized but low-diversity system of intertidal crustaceans. This serves as a pilot for far more diverse systems.

#1. You have a bunch of FASTA files of sequence data, each file is a population. Read them in with loop (or WHILE), so each has an index number, each population has a short name.

#2. Each population run summary statistics and attach those to the short name (array)

#3. All sequences in 'unknown' file are BLASTed to all populations. Max hit (best match) is counted for each; if best hit is still x% divergent, they are put into 'unknown' category, otherwise they are pooled into category of best hit.

#4. unknowns distance matrix, cluster and break into K populations somehow appropriate, perhaps by x'% divergent, label as unknown1...unknownZ

#5. unknowns BLAST to known, net nucleotide divergence to nearest known is recorded along with that KNOWN

#6. now you have all populations characterized and named either by what they BLAST to or what they are closest to (and how distant)

#7. summary statistics on all the populations from the sample

```{r ZOMG more params}
#A. actually, the within-species (population) variance in volume and differential PCR less problem so 
#   you do actually have some haplotype frequency data and site frequency spectrum data?? Hmm not quite. because deep coverage of each allele but unknown coverage, so SFS would be goofy.

# 
#B. you do have biomass and/or number individuals, e.g. Marcus could certainly tell me if n=200 in the library
# putting 20, 200, 2000 individuals on HiSeq to barcode result in ~ same proportion of reads
# (but ridiculous variance in depth of coverage among those runs for same data)
# BUT this info helps constrain/tune the POSTERIOR that is in units of individuals rather than proportions

maxindivs=250
```


```{r NotoPrior, echo=FALSE,warning=FALSE,results='hide',message=FALSE,fig.show='asis'}
# rough estimates from Zakas Q is 10, hapdiv is 0.7
# grabbing a file from Geneious....
library(PopGenome)
file<-readData("FastaSeqs") # actual sample size here is 20. That is the number I'd like to come out...AND, TO AN EXTENT, IT IS AT LEAST IN THE INTERVALS FOR BOTH...
actual<-20
file@n.sites
basic<-diversity.stats(file)
Hapdiv<-basic@hap.diversity.within

Nsite<-basic@n.biallelic.sites

#hapcount is not reporting back right now.
basic2<-detail.stats(file)
basic3<-Achaz.stats(file)
numhaps<-basic2@haplotype.counts
#that doesn't work either.
numhaps=10 #for THIS SMALL DATA SET
Hapdiv #haplotype diversity
Nsite #number of seg sites taht are biallelic, for now assume ISM

```


# do we want to use this script to re-do my old code for grabbing data from Genbank adn getting sumstats? now could use the package in Jehlius/picoroco papers...

#8. we use summary statistics from sample populations relative to baseline PRIOR to estimate a POSTERIOR distribution of abundance in that sample.

```{r hapdiv, echo=FALSE,warning=FALSE,results='hide',message=FALSE,fig.show='asis'}

####################################
#numhaps=9
#hapdiv=0.7 #minvalue 0.001 if no information; this is PRIOR information
####################################

x=1
cdf=0
indprob=0
array<-NULL
while (cdf<0.99) {
  cdfprev<-cdf
#  cdf<-pgamma(x,1,Hapdiv) 
#  if use 1 as shape parameter keeping shape parameter constant doesn't account for increased variance (?) as numhaps go up, 
#  e.g. error may be higher as you observe more...once it is working run it by somebody mathier.
  cdf<-pgamma(x,numhaps,Hapdiv) #might be that numhaps is actually the shape parameter!!!! or: something else...non-gamma.
  indprob<-cdf-cdfprev
  
  
  happrob<-numhaps+(x-1)
  array<-c(array,happrob)
  array<-c(array,cdf)
  array<-c(array,indprob)
#  print(happrob)
#  print(cdf)
  x=x+1
}

par(mfrow=c(3,1))

probs<-t(matrix(array,nrow=3))
probs
#plot(probs[,1],probs[,3],col='red')

```



```{r WakeleyCh4, echo=FALSE,warning=FALSE,results='hide',message=FALSE,fig.show='asis'}
library(plot3D)
Q = 2 # this is PRIOR information
maxn = min(5*Q,maxindivs) #above 70 this behaves funny???? OR IT MAY BE SOME MULTIPLE OF Q*maxn that is problem?

obsvdk = Nsite
maxk = 2*obsvdk #must be greater than obsvdk



a <- c(1:maxn)
b <- c(1:maxk)
PSKv <- numeric(maxn*maxk)
PSKv <- matrix(PSKv,ncol=maxn)
colnames(PSKv)<-a
rownames(PSKv)<-b

for (n in 2:maxn) {
#  print ("n")
#  print (n)
  for (k in 0:maxk) {
#    print (k)
    PSK=0
    for (i in 2:n) {
#      print (i)
      PSK<-PSK + ((-1)^i)*(choose((n-1),(i-1)))*((i-1)/(Q+i-1))*(Q/(Q+i-1))^k
      
    }
#    print (PSK)
    PSKv[k,n]<-PSKv[k,n]+PSK
    obsvd<-PSKv[obsvdk,]

  }
  
}

hist3D(x=seq(0,1,length.out=nrow(PSKv)),y=seq(0,1,length.out=ncol(PSKv)),PSKv,col=NULL,border="black",theta=30,phi=30,xlab="k",ylab="n",zlab="P{S=k}",alpha=0.35,main=paste((expression(theta))," = ",Q))

```

```{r nextfig, echo=FALSE,warning=FALSE,results='hide',message=FALSE,fig.show='asis'}
#  print(obsvd) #now would be easy to print 95% HPD of obsvd, so for given theta and k this is the distribution of n
  plot(obsvd,xlab="n",ylab="P(n|k)",ylim=c(0,0.15),main=paste("given segsites =",obsvdk, "and theta =",Q,"in black; given hapdiv in red"))
points(probs[,1],probs[,3],col='red')
abline(v=actual)


```



#9. can choose to even update the PRIOR based on this posterior, e.g. unknown 1 is now a population with SOME information on summary statistics (depends on sample size)

#10. relative species abundance curve, most common species is in position 1 on X axis and abundance is plotted on Y as boxplot of likely abundance.... all the way down to the populations that you can only tell are present

#11. you are now in the lognormal world of Hubbell, flawed though that may be.

