---
title: "Simulations"
author: "Paula Pappalardo"
date: "Friday, April 10, 2015"
output: pdf_document
---


```{r testing how to simulate data, eval=FALSE}
# Here we have the detailed explanation of how to simulate data, I wrote a little bit of how to do it in PopGenome but I'm using here the approach in the phyloclust package that allows to work in windows too.
# using PopGenome----
# library(PopGenome)
# PopGenome has the MS function that allows to use for coalescent simulations to derive the expected neutral distribution of Tajima's D. 
# better help for the MS function
# http://www.inside-r.org/packages/cran/PopGenome/docs/MS
# The problem with this is that you need to install the C files of the ms program, and that cannot be done in windows. In this link you can find ms program http://home.uchicago.edu/rhudson1/source/mksamples.html
# The following code should work if that is achieve in linux or macs.
# load data
#file<-readData("FastaSeqs")
# we apply the neutrality stats and with that output we run simulations
#file<-neutrality.stats(file,list(1:6))
# we call Hudson program
#ms.file <- MS(file, thetaID="Tajima", neutrality=TRUE)
# we can extract the simulated Tajima's D values
#MS.get.stats(ms.file)

# Another option is to use phyclust package that has a ms function included----
library(phyclust)
# http://thirteen-01.stat.iastate.edu/snoweye/phyclust/?item=document

# To read sequences from a fasta file
data<-read.fasta("FastaSeqs/NotoSample.fas")
class(data) # class 'seq.data'
data$org.code # this is the matrix with our original data
data$nseq # number of sequences in our data
data$seqlen # length of the sequences

# we need to set a seed for our simulations
set.seed(1234)

# to simulate trees (we use our number of sequences as number of tips)
data.trees<-ms(nsam=data$nseq, nreps=1, opts= "-T -t 0.2")

# to extract the tree and convert it to 'phylo' object
simtree<-read.tree(text=data.trees[3])
simtree

# to simulate sequences (we use both our sequence length and number of sequences)
set.seed(1234)
# -l is to set the lenght of sequences, sadly, it only accepts the number and we cannot use data$seqlen or a variable (which would be more useful for us for loops)
data.seqs<-seqgen(opts= "-mHKY -l481 -q", rooted.tree=simtree)
class(data.seqs) #seqgen

# to read the 'seqgen' object and convert to 'seq.data'
data.seqs.easy<-read.seqgen(data.seqs)
class(data.seqs.easy) #seq.data
str(data.seqs.easy) # in $org we can find the sequences

# to save in fasta file so we can use with PopGenome
write.fasta(data.seqs.easy$org,"simSequences.fas")

```

```{r creating datasets of different size from picoroco file}
library(ape)
library(muscle)
library(seqinr)

# I used all this chunk at the beggining, to read sequences and convert to fasta, and check alignment, but the final files used for the analysis were edited by eye in MEGA, so I just load those in next chunks

# We need to convert the "Sequence" column to characters
# to check all the samples
austroGB$SequenceNew<-as.character(austroGB$Sequence)
austroSeq <- strsplit(austroGB$SequenceNew, "")
names(austroSeq)<-austroGB$Label

# to use just the picorocos
picoroco$SequenceNew<-as.character(picoroco$Sequence)
picoSeq <- strsplit(picoroco$SequenceNew, "") #is alist, each component is a list of characters with all the bases
names(picoSeq)<-picoroco$Label # to add the labels

# to transform to a FASTA file and use it with the "muscle" package
seqinr::write.fasta(austroSeq, names=austroGB$Label,file.out = "allSamples.fasta")
# to transform to a FASTA file and use it with the "muscle" package
seqinr::write.fasta(picoSeq, names=picoroco$Label,file.out = "pico.fasta")

# now we can align the sequences with the function muscle of the muscle package, if out=NULL it returns a muscle object, but if it is not NULL, it returns a FASTA file in the folder indicated
# this one is to use later
muscle::muscle("pico.fasta",out="C:/Users/Paula/Dropbox/Picoroco MS/alignout/picoroco.fasta")

```



```{r simulating diversity indexes with PopGenome}
library(PopGenome)
# PopGenome has the MS function that allows to use for coalescent simulations to derive the expected neutral distribution of Tajima's D. 
# better help for the MS function
# http://www.inside-r.org/packages/cran/PopGenome/docs/MS

# load initial data
file<-readData("FastaSeqs",include.unknown=T)
get.sum.data(file)

# first we need to apply the neutrality stats
file<-neutrality.stats(file)

# or the F_ST stats if we want haplotype diversity
file<-F_ST.stats(file)
file@Pi #Nei's diversity (within population)
file@region.stats@haplotype.counts->haplotypecounts

# to get number of haplotypes
hapfreq<-unlist(haplotypecounts)
actual<-sum(hapfreq)
hapfreq<-hapfreq/actual
numhaps<-length(hapfreq)

# to get Pi in another way
diversity.stats(file,pi=TRUE)->file
file@Pi #also says is Nei's pi

# we call Hudson program, thetaID needs to be defined for any module
ms.file <- MS(file, niter=20,thetaID="Tajima", neutrality=TRUE)
ms.file

ms.file <- MS(file, niter=20,thetaID="Tajima", F_ST=TRUE)
ms.file

# we can extract the simulated Tajima's D (and other stats as well) in form of a matrix
MS_getStats(ms.file) #stats change depending what module we choose

# we can extract the pvalue of simulated values <= than observed
ms.file@prob.less[1]
# also the observed values for each test
ms.file@obs.val[1]

# To define PARAMETERS we need to create an object of class "test.params" and use the option params=file in the MS function
    # to create the object
    parDefined<-new("test.params")
    
    # to add values #http://www.inside-r.org/node/132342 
    # ---theta---- it has to be a vector of lenght n.regions
    parDefined@theta<-5
    # ----growth values-----if (+) means that the population was smaller in the past, if (-) the population was larger in the past
    # a growth value of 10 is the ones that minimize Tajima's D
    parDefined@growth<-10
    
    # ----segregating sites-----single numeric value
    #parDefined@fixed.seg.sites<-1
    
    # ----seeds----specify 3 random number seeds. a vector of length 3 with positive integer values is expected.IS NOT WORKING, ERROR MESSAGE
    #parDefined@seeds<-c(1,2,3)

#to run simulations and get stats
    set.seed(123)
    ms.file1<- MS(file,niter=30, thetaID="user", params=parDefined,neutrality=TRUE)
    set.seed(123)
    ms.file2<- MS(file, niter=10, thetaID="user", params=parDefined,F_ST=TRUE)
    MS_getStats(ms.file1)->sumFile1
    MS_getStats(ms.file2)->sumFile2
    sumFile1
    View(sumFile2)

theta= 2.005128
a=NULL
for (i in 1:20){
  (theta/(theta+i-1))->a[i+0]
  }
sum(a)

# to average estimators accross all the simulated populations
# neutrality module
mean(sumFile1[,"Tajima.D"])
mean(sumFile1[,"theta_Tajima"])
var(sumFile1[,"theta_Tajima"])
mean(sumFile1[,"theta_Watterson"])
var(sumFile1[,"theta_Watterson"])

#MSMS allows for simulation with selection, we need the folder with the program and executable files to run it
http://www.mabs.at/ewing/msms/quickStart.shtml
```


We simulated sequences data for 200 individuals using the combined approach ms [@Hudson2002] plus seq-gen [@Rambaut1997] implemented in the *phyclust* package [@Chen2011] of the R program [@RCoreTeam2013]. The ms() function is used to simulate trees in a population that evolves under a Wrigth-Fisher model, and the seqgen() function is used to generate the nucleotide sequences according to those/the tree/s. We simulated sequences for different levels of mutation rate ($\theta$) and saved the  sequences in fasta files representing our "original" populations. 


```{r does not work but just in case,eval=FALSE}
library(phyclust)

# We are going to simulate sequence data for 200 individuals using different thetas. I tried with a loop already but the problem is that ms and seq-gen don't accept variables within their arguments.

# simulating trees----
set.seed(123); tree1<-ms(nsam=20, opts= "-T -t 1")
simtree1<-read.tree(text=tree1[3])
set.seed(123); tree2<-ms(nsam=20, nreps=1, opts= "-T -t 2")
simtree2<-read.tree(text=tree2[3])
set.seed(123); tree3<-ms(nsam=20, nreps=1, opts= "-T -t 3")
simtree3<-read.tree(text=tree3[3])
set.seed(123); tree4<-ms(nsam=20, nreps=1, opts= "-T -t 4")
simtree4<-read.tree(text=tree4[3])
set.seed(123); tree5<-ms(nsam=200, nreps=1, opts= "-T -t 5")
simtree5<-read.tree(text=tree5[3])
set.seed(123); tree6<-ms(nsam=200, nreps=1, opts= "-T -t 6")
simtree6<-read.tree(text=tree6[3])
set.seed(123); tree7<-ms(nsam=200, nreps=1, opts= "-T -t 7")
simtree7<-read.tree(text=tree7[3])
set.seed(123); tree8<-ms(nsam=200, nreps=1, opts= "-T -t 8")
simtree8<-read.tree(text=tree8[3])
set.seed(123); tree9<-ms(nsam=200, nreps=1, opts= "-T -t 9")
simtree9<-read.tree(text=tree9[3])
set.seed(123); tree10<-ms(nsam=20, opts= "-T -t 10")
simtree10<-read.tree(text=tree10[3])

# generating sequences from those trees----
set.seed(123); seqs1<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree1)
seqs.easy1<-read.seqgen(seqs1)
set.seed(123); seqs2<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree2)
seqs.easy2<-read.seqgen(seqs2)
set.seed(123); seqs3<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree3)
seqs.easy3<-read.seqgen(seqs3)
set.seed(123); seqs4<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree4)
seqs.easy4<-read.seqgen(seqs4)
set.seed(123); seqs5<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree5)
seqs.easy5<-read.seqgen(seqs5)
set.seed(123); seqs6<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree6)
seqs.easy6<-read.seqgen(seqs6)
set.seed(123); seqs7<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree7)
seqs.easy7<-read.seqgen(seqs7)
set.seed(123); seqs8<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree8)
seqs.easy8<-read.seqgen(seqs8)
set.seed(123); seqs9<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree9)
seqs.easy9<-read.seqgen(seqs9)
set.seed(123); seqs10<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree10)
seqs.easy10<-read.seqgen(seqs10)

# writing sequences to FASTA files----
write.fasta(seqs.easy1$org,"simPopTheta1.fas")
write.fasta(seqs.easy2$org,"simPopTheta2.fas")
write.fasta(seqs.easy3$org,"simPopTheta3.fas")
write.fasta(seqs.easy4$org,"simPopTheta4.fas")
write.fasta(seqs.easy5$org,"simPopTheta5.fas")
write.fasta(seqs.easy6$org,"simPopTheta6.fas")
write.fasta(seqs.easy7$org,"simPopTheta7.fas")
write.fasta(seqs.easy8$org,"simPopTheta8.fas")
write.fasta(seqs.easy9$org,"simPopTheta9.fas")
write.fasta(seqs.easy10$org,"simPopTheta10.fas")

```

```{r failed loop}
# We are going to simulate sequence data for 200 individuals using different thetas:
# Define theta values to use, for now I put from 1 to 10
theta<-c(1:10)

# Loop in theta, simulate sequences and save in fasta file
for (i in theta){
    print(paste('Working in theta',i))
    # first, simulate tree
      set.seed(1234)
      tree<-ms(nsam=20, nreps=1, opts= "-T -t i")
      simtree<-read.tree(text=tree[3])
    # second, generate 200 sequences with a length of 600 bp
      set.seed(1234)
      seqs<-seqgen(opts= "-mHKY -l600 -q", rooted.tree=simtree)
    # third, read and save the simulated sequences
      seqs.easy<-read.seqgen(seqs)
      filename<-paste("simSeq",i,".fas",sep="")
      write.fasta(seqs.easy$org,filename)
    }

# Ok, the loop looks nice but it doesn't work, the call of ms doesn't allow a variable, you need to write the number. This can probably be done with the "tbs" calls of the ms program, but I need to learn more on that, what I tried crush R...
```

```{r simulating datasets with phyclust,eval=FALSE}
library(phyclust)

# trying to think because is still making similar sequences!
set.seed(123)
tree<-ms(nsam=20, nreps=1, opts= "-T -t 1 -G 10")
simtree<-read.tree(text=tree[3])
set.seed(123)

tree2<-ms(nsam=20, nreps=1, opts= "-T -t 10")
simtree2<-read.tree(text=tree2[3])
par(mfrow = c(1, 2))
plot(simtree1,main="Theta 1")
plot(simtree2,main="Theta 10")

# changing -s, according to the ms manual, changes theta
set.seed(123)
seqs1<-seqgen(opts= "-mHKY -l600 -q -s0.004", rooted.tree=simtree)
seqs.easy1<-read.seqgen(seqs1)
write.fasta(seqs.easy1$org,"test_s0004.fas")
set.seed(123)
seqs2<-seqgen(opts= "-mHKY -l600 -q -s0.0035", rooted.tree=simtree)
seqs.easy2<-read.seqgen(seqs2)
write.fasta(seqs.easy2$org,"test_s00035.fas")
set.seed(123)
seqs3<-seqgen(opts= "-mHKY -l600 -q -s0.003", rooted.tree=simtree)
seqs.easy3<-read.seqgen(seqs3)
write.fasta(seqs.easy3$org,"test_s0003.fas")

# sequences are identical!!#$%& sigh...

# a1 for my example with 20 samples
a1<-1+1/2+1/3+1/4+1/5+1/6+1/7+1/8+1/9+1/10+1/11+1/12+1/13+1/14+1/15+1/16+1/17+1/18+1/19

```

For each of the simulated populations (200 individuals each) we estimated the genetic diversity using different indexes included in the package *PopGenome*. Subsequently, we sampled populations of 20 individuals to simulate our observations, and used backcalculations to estimate the original population size.

```{r genetic diversity in sim populations}
library(PopGenome) 

# In PopGenome, we can read all our fasta files at the same time, from our defined folder
# I think I found a useful trick to name the path that will work for both of us
ourpath<-getwd()
gitpath<-paste(ourpath,"/SimSeqs",sep="")

# converting FASTA file to GENOME object
simSeq<-readData(gitpath,include.unknown=T)
simSeq@n.sites # gives the number of sites in the alignment
get.sum.data(simSeq) # gives the summary information of the alignment, with the @ you can get all of those summaries, for example:
Varsites<-simSeq@n.biallelic.sites #number of biallelic sites (SNPs)
Varsites

# aplying the "diversity.stats" methods
diversity.stats(simSeq,pi=TRUE)->divResults

Piout<-divResults@Pi
Piout
Totsites<-divResults@n.sites
Piout2<-Piout/Totsites #Pi per nucleotide
# Haplotype diversity (H)
divResults@hap.diversity.within

# aplying the "neutrality.stats"
neuResults<-neutrality.stats(simSeq)
neu<-get.neutrality(neuResults,theta=T,stats=T)

Tajout<-neuResults@Tajima.D
Tajout
neuResults@n.segregating.sites

```

```{r subsamples}
# We are going to "sample" 20 individuals from our simulated populations
sizeNorth<-length(northLab)
sizeSouth<-length(southLab)

Pop1<-readData("C:/Users/Paula/Dropbox/Picoroco MS/alignout", include.unknown=T)
ranSnnDivline = NULL

for (i in 1:reps) {
  # we sample from the pico file to make random northern group
  randnorth<-as.character(sample(pico$Label,sizeNorth,replace=F))
  # we keep the ones not sampled to make the southern group
  randsouth<-as.character(setdiff(pico$Label, randnorth))
  # now we set the random sampled populations for analysis
  ranResults<-set.populations(ranPico,list(randnorth,randsouth))
  ranResults@region.data@populations2 #check it works
  ranSnn<-F_ST.stats.2(ranResults,snn=TRUE)
  # and we saved the results for each rep in the ranSnnout vector
  ranSnnDivline[i]<-ranSnn@Hudson.Snn
}
SnnquantDivline<-quantile(ranSnnDivline,probs=c(0.05,0.95))
SnnPdivline<-ecdf(ranSnnDivline)
SnnPvalDivline<-1-SnnPdivline(SnnByDivline)
```

